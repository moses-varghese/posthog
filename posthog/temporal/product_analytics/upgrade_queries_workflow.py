import dataclasses
import datetime as dt

from django.db import connection
import temporalio.workflow
from posthog.models import Insight
from posthog.schema_migrations.upgrade import upgrade
from posthog.temporal.common.base import PostHogWorkflow
from posthog.temporal.common.logger import get_internal_logger
from posthog.schema_migrations import LATEST_VERSIONS


@temporalio.activity.defn
def get_insights_to_migrate() -> list[int]:
    # TODO: Add index
    # TODO: Cross-join or run separately for each kind?
    # CREATE INDEX insight_query_gin_path
    # ON insight USING gin (query jsonb_path_ops);
    latest_versions = []
    for kind, version in LATEST_VERSIONS.items():
        latest_versions.append(f"('{kind}', {version})")

    sql = f"""
        WITH latest(kind, v_latest) AS (
            VALUES
                {','.join(latest_versions)}
        )
        SELECT DISTINCT i.id
        FROM posthog_dashboarditem AS i
        CROSS JOIN latest AS l
        WHERE jsonb_path_exists(
            i.query,
            format(
                $$
                $.** ? (
                    @.kind == "%s" &&
                    ( @.v < %s || @.v == null )
                )
                $$,
                l.kind,
                l.v_latest
            )::jsonpath
        );
    """

    with connection.cursor() as cur:
        cur.execute(sql)
        ids = [row[0] for row in cur.fetchall()]

    return ids


@dataclasses.dataclass(frozen=True)
class MigrateInsightsBatchInputs:
    """Inputs for the migrate insights batch activity."""

    insight_ids: list[int] = dataclasses.field()


@temporalio.activity.defn
def migrate_insights_batch(inputs: MigrateInsightsBatchInputs) -> None:
    """Migrate a batch of insights to the latest version."""
    logger = get_internal_logger()

    insights = Insight.objects.filter(id__in=inputs.insight_ids)

    for insight in insights:
        try:
            insight.query = upgrade(insight.query)
            insight.save()
        except Exception as e:
            logger.exception(f"Error migrating insight {insight.id}: {str(e)}")


@temporalio.workflow.defn(name="upgrade-queries")
class UpgradeQueriesWorkflow(PostHogWorkflow):
    @temporalio.workflow.run
    async def run(self):
        # TODO: use a while-loop to process insights in batches
        insight_ids = await temporalio.workflow.execute_activity(
            get_insights_to_migrate,
            None,
            start_to_close_timeout=dt.timedelta(minutes=5),
            retry_policy=temporalio.common.RetryPolicy(
                initial_interval=dt.timedelta(minutes=5),
                maximum_interval=dt.timedelta(minutes=60),
                maximum_attempts=3,
            ),
        )

        for i in range(0, len(insight_ids)):
            batch = insight_ids[i : i + 100]

            temporalio.workflow.execute_activity(
                migrate_insights_batch,
                MigrateInsightsBatchInputs(insight_ids=[insight.id for insight in batch]),
                start_to_close_timeout=dt.timedelta(minutes=10),
                retry_policy=temporalio.common.RetryPolicy(
                    initial_interval=dt.timedelta(minutes=10),
                    maximum_interval=dt.timedelta(minutes=60),
                    maximum_attempts=3,
                ),
            )
