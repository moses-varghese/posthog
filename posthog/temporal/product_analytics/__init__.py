from .upgrade_queries_workflow import (
    UpgradeQueriesWorkflow,
    get_insights_to_migrate,
    migrate_insights_batch,
)

WORKFLOWS = [
    UpgradeQueriesWorkflow,
]
ACTIVITIES = [
    get_insights_to_migrate,
    migrate_insights_batch,
]
