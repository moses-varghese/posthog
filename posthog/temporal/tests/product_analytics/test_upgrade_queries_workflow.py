import pytest

from posthog.models.insight import Insight
from posthog.schema import NodeKind
from posthog.schema_migrations import LATEST_VERSIONS
from posthog.temporal.product_analytics.upgrade_queries_workflow import get_insights_to_migrate
from posthog.test.base import QueryMatchingTest


@pytest.fixture(autouse=True)
def setup_migrations():
    LATEST_VERSIONS.clear()
    # MIGRATIONS.clear()

    # MIGRATIONS[NodeKind.TRENDS_QUERY] = {1: SampleMigration()}
    # MIGRATIONS[NodeKind.EVENTS_NODE] = {1: EventsNodeMigration()}
    LATEST_VERSIONS[NodeKind.INSIGHT_VIZ_NODE] = 3
    LATEST_VERSIONS[NodeKind.TRENDS_QUERY] = 5
    LATEST_VERSIONS[NodeKind.EVENTS_NODE] = 7

    yield


class TestUpgradeQueriesWorkflow(QueryMatchingTest):
    @pytest.mark.django_db(transaction=True)
    def test_get_insights_to_migrate(self, activity_environment, team):
        # all versions satisfied
        Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                    "v": 6,
                },
                "v": 4,
            },
            team=team,
        )

        # matching top-level node (InsightVizNode)
        m1 = Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                    "v": 6,
                },
                "v": 3,
            },
            team=team,
        )

        # matching node in nested dict (TrendsQuery)
        m2 = Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                    "v": 5,
                },
                "v": 4,
            },
            team=team,
        )

        # matching node in nested list (EventsNode)
        m3 = Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 7}],
                    "v": 6,
                },
                "v": 4,
            },
            team=team,
        )

        # no query
        Insight.objects.create(team=team)

        # kind without migration (DataTableNode)
        Insight.objects.create(
            query={
                "kind": "DataTableNode",
            },
            team=team,
        )

        # no version (InsightVizNode)
        m4 = Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                    "v": 6,
                },
            },
            team=team,
        )

        # none version (InsightVizNode)
        m5 = Insight.objects.create(
            query={
                "kind": "InsightVizNode",
                "source": {
                    "kind": "TrendsQuery",
                    "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                    "v": 6,
                },
                "v": None,
            },
            team=team,
        )

        # with snapshot_postgres_queries_context(self):
        result = activity_environment.run(get_insights_to_migrate)

        expected_ids = [m1.id, m2.id, m3.id, m4.id, m5.id]
        assert sorted(result), expected_ids
