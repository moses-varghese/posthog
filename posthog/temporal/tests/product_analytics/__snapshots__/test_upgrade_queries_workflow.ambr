# serializer version: 1
# name: TestUpgradeQueriesWorkflow.test_get_insights_to_migrate
  '''
  WITH latest(kind, v_latest) AS (
                                  VALUES ('InsightVizNode',
                                          3),('TrendsQuery',
                                              5),('EventsNode',
                                                  7))
  SELECT DISTINCT i.id
  FROM posthog_dashboarditem AS i
  CROSS JOIN latest AS l
  WHERE jsonb_path_exists(i.query, format($$
                  $.** ? (
                      @.kind == "%s" &&
                      ( @.v < %s || @.v == null )
                  )
                  $$, l.kind, l.v_latest)::jsonpath);
  '''
# ---
